array_api_strict/tests/test_array_object.py

```diff
@@ -410,6 +411,78 @@ def _matmul_array_vals():
                     x.__imatmul__(y)
 
 
+@pytest.mark.parametrize(
+    "op",
+    [
+        op for op, dtypes in binary_op_dtypes.items()
+        if dtypes not in ("real numeric", "floating-point")
+    ],
+)
+def test_binary_operators_vs_numpy_int(op):
+    """np.int64 is not a subclass of int and must be disallowed"""
+    a = asarray(1)
+    i64 = np.int64(1)
+    with pytest.raises(TypeError, match="Expected Array or Python scalar"):
+        getattr(a, op)(i64)
+
+
+@pytest.mark.parametrize(
+    "op",
+    [
+        op for op, dtypes in binary_op_dtypes.items() 
+        if dtypes not in ("integer", "integer or boolean")
+    ],
+)
+def test_binary_operators_vs_numpy_float(op):
+    """
+    np.float64 is a subclass of float and must be allowed.
+    np.float32 is not and must be rejected.
+    """
+    a = asarray(1.)
+    f64 = np.float64(1.)
+    f32 = np.float32(1.)
+    func = getattr(a, op)
+    for op in binary_op_dtypes:
+        assert isinstance(func(f64), Array)
+        with pytest.raises(TypeError, match="Expected Array or Python scalar"):
+            func(f32)
+
+
+@pytest.mark.parametrize(
+    "op",
+    [
+        op for op, dtypes in binary_op_dtypes.items()
+        if dtypes not in ("integer", "integer or boolean", "real numeric")
+    ],
+)
+def test_binary_operators_vs_numpy_complex(op):
+    """
+    np.complex128 is a subclass of complex and must be allowed.
+    np.complex64 is not and must be rejected.
+    """
+    a = asarray(1.)
+    c64 = np.complex64(1.)
+    c128 = np.complex128(1.)
+    func = getattr(a, op)
+    for op in binary_op_dtypes:
+        assert isinstance(func(c128), Array)
+        with pytest.raises(TypeError, match="Expected Array or Python scalar"):
+            func(c64)
+
+
+@pytest.mark.parametrize("op,dtypes", binary_op_dtypes.items())
+def test_binary_operators_device_mismatch(op, dtypes):
+    if dtypes in ("real numeric", "floating-point"):
+        dtype = float64
+    else:
+        dtype = int64
+    
+    a = asarray(1, dtype=dtype, device=CPU_DEVICE)
+    b = asarray(1, dtype=dtype, device=Device("device1"))
+    with pytest.raises(ValueError, match="different devices"):
+        getattr(a, op)(b)
```

Comment:
This test would be great to extend to binary functions, too. I'd expect them to work just the same, because they reuse the `._promote_scalar` logic. If `xp.add(a, b)` differs from `a.__add__(b)`, it's a bug, so would be great to test for.

Reply:
It turns out extending to binary functions caused the scope blow up, so I moved it to a separate PR: #148